File: java/DeleteAndEarn.java
import java.util.*;
import java.util.stream.Collectors;


public class DeleteAndEarn {
    public int deleteAndEarn(int [] nums) {
        Map<Integer, Integer> numberCount = new HashMap<>();

        for(int num: nums) {
            int count = numberCount.getOrDefault(num, 0);
            numberCount.put(num, num + count);
        }

        List<Integer> keys = numberCount.keySet().stream().sorted(Integer::compareTo).collect(Collectors.toList());
        int prevPrevValue = 0, prevValue = 0, value = 0;
        int prevKey = -1;

        for(int key: keys) {
            value = numberCount.get(key);
            if(prevKey != -1) {
                if (key - prevKey == 1) {
                    value = Math.max(prevPrevValue + value, prevValue);
                } else {
                    value = value + prevValue;
                }
            }
            prevPrevValue = prevValue;
            prevValue = value;
            prevKey = key;
        }

        return value;
    }

    public static void main(String[] args) {
        DeleteAndEarn deleteAndEarn = new DeleteAndEarn();
        int result = deleteAndEarn.deleteAndEarn(new int[]{3, 4, 2});
        System.out.println(result);
    }
}

File: python/delete_and_earn.py
from typing import List
from collections import defaultdict


class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        sub_sum = defaultdict(int)
        start_value = 10000
        end_value = 1
        for num in nums:
            sub_sum[num] += num
            if start_value > num:
                start_value = num
            if end_value < num:
                end_value = num

        for i in range(start_value, end_value + 1):
            sub_sum[i] = max(sub_sum[i] + sub_sum[i - 2], sub_sum[i - 1])

        return sub_sum[end_value]


s = Solution()
print(s.deleteAndEarn([3,4,2]))
print(s.deleteAndEarn([2,2,3,3,3,4]))
print(s.deleteAndEarn([1]))
print(s.deleteAndEarn([1, 1, 1, 2, 3]))